File: src/main/java/com/noximity/remmyChat/RemmyChatPlaceholders.java
package com.noximity.remmyChat;

import com.noximity.remmyChat.models.ChatUser;
import me.clip.placeholderapi.expansion.PlaceholderExpansion;
import org.bukkit.OfflinePlayer;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public class RemmyChatPlaceholders extends PlaceholderExpansion {

    private final RemmyChat plugin;

    public RemmyChatPlaceholders(RemmyChat plugin) {
        this.plugin = plugin;
    }

    @Override
    public @NotNull String getIdentifier() {
        return "remmychat";
    }

    @Override
    public @NotNull String getAuthor() {
        return String.join(", ", plugin.getDescription().getAuthors());
    }

    @Override
    public @NotNull String getVersion() {
        return plugin.getDescription().getVersion();
    }

    @Override
    public boolean persist() {
        return true; // This is required or else PlaceholderAPI will unregister the expansion on reload
    }

    @Override
    public @Nullable String onRequest(OfflinePlayer player, @NotNull String params) {
        if (player == null) {
            return "";
        }

        if (params.equalsIgnoreCase("msgtoggle")) {
            ChatUser user = plugin.getChatService().getChatUser(player.getUniqueId());
            return String.valueOf(user.isMsgToggle());
        }

        if (params.equalsIgnoreCase("socialspy")) {
            ChatUser user = plugin.getChatService().getChatUser(player.getUniqueId());
            return String.valueOf(user.isSocialSpy());
        }

        if (params.equalsIgnoreCase("channel")) {
            if (player.isOnline()) {
                ChatUser user = plugin.getChatService().getChatUser(player.getUniqueId());
                return user.getCurrentChannel();
            }
            return plugin.getConfigManager().getDefaultChannel().getName();
        }

        if (params.equalsIgnoreCase("group") && plugin.getPermissionService().isLuckPermsHooked()) {
            if (player.isOnline()) {
                return plugin.getPermissionService().getPrimaryGroup(player.getPlayer());
            }
            return "";
        }

        return null;
    }
}

File: bin/main/com/noximity/remmyChat/RemmyChatPlaceholders.kt
package com.noximity.remmyChat

import me.clip.placeholderapi.expansion.PlaceholderExpansion
import org.bukkit.OfflinePlayer

class RemmyChatPlaceholders(private val plugin: RemmyChat) : PlaceholderExpansion() {

    override fun getIdentifier(): String {
        return "remmychat"
    }

    override fun getAuthor(): String {
        return plugin.description.authors.joinToString(", ")
    }

    override fun getVersion(): String {
        return plugin.description.version
    }

    override fun persist(): Boolean {
        return true // This is required or else PlaceholderAPI will unregister the expansion on reload
    }

    override fun onRequest(player: OfflinePlayer?, params: String): String? {
        if (player == null) {
            return ""
        }

        when (params.lowercase()) {
            "msgtoggle" -> {
                val user = plugin.chatService.getChatUser(player.uniqueId) ?: return ""
                return user.isMsgToggle.toString()
            }

            "socialspy" -> {
                val user = plugin.chatService.getChatUser(player.uniqueId) ?: return ""
                return user.isSocialSpy.toString()
            }

            "channel" -> {
                if (player.isOnline) {
                    val user = plugin.chatService.getChatUser(player.uniqueId) ?: return ""
                    return user.currentChannel
                }
                return plugin.configManager.defaultChannel?.name ?: ""
            }

            "group" -> {
                if (plugin.permissionService.isLuckPermsHooked && player.isOnline) {
                    val onlinePlayer = player.player
                    if (onlinePlayer != null) {
                        return plugin.permissionService.getPrimaryGroup(onlinePlayer) ?: ""
                    }
                }
                return ""
            }
        }

        return null
    }
}
